diff --git a/node_modules/country-state-city/lib/city.js b/node_modules/country-state-city/lib/city.js
index e09ddbb..c447697 100644
--- a/node_modules/country-state-city/lib/city.js
+++ b/node_modules/country-state-city/lib/city.js
@@ -1,4 +1,4 @@
-import cityList from './assets/city.json';
+const cityList = require('./assets/city.json')
 import { compare } from './utils';
 // Get a list of all cities.
 function getAllCities() {
diff --git a/node_modules/country-state-city/lib/country.js b/node_modules/country-state-city/lib/country.js
index f90e4f6..31a8399 100644
--- a/node_modules/country-state-city/lib/country.js
+++ b/node_modules/country-state-city/lib/country.js
@@ -1,4 +1,5 @@
-import countryList from './assets/country.json';
+const countryList = require('./assets/country.json')
+
 import { findEntryByCode } from './utils';
 // Get a country by isoCode.
 function getCountryByCode(isoCode) {
diff --git a/node_modules/country-state-city/lib/state.js b/node_modules/country-state-city/lib/state.js
index ab326f8..959bdd5 100644
--- a/node_modules/country-state-city/lib/state.js
+++ b/node_modules/country-state-city/lib/state.js
@@ -1,4 +1,4 @@
-import stateList from './assets/state.json';
+const stateList = require('./assets/state.json')
 import { findEntryByCode, findStateByCodeAndCountryCode, compare } from './utils';
 // Get a list of all states.
 export function getAllStates() {
