// @apollo/client/cache/inmemory/helpers.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../../utilities/globals/index.js');
var directives = require('../../utilities/graphql/directives.js');
var fragments = require('../../utilities/graphql/fragments.js');
var getFromAST = require('../../utilities/graphql/getFromAST.js');
var storeUtils = require('../../utilities/graphql/storeUtils.js');
require('../../utilities/graphql/transform.js');
var mergeDeep = require('../../utilities/common/mergeDeep.js');
require('../../utilities/observables/Observable.js');
var objects = require('../../utilities/common/objects.js');
require('../../utilities/observables/Concast.js');
require('../../utilities/common/canUse.js');
var compact = require('../../utilities/common/compact.js');

var hasOwn = Object.prototype.hasOwnProperty;
function isNullish(value) {
    return value === null || value === void 0;
}
var isArray = Array.isArray;
function defaultDataIdFromObject(_a, context) {
    var __typename = _a.__typename, id = _a.id, _id = _a._id;
    if (typeof __typename === "string") {
        if (context) {
            context.keyObject =
                !isNullish(id) ? { id: id } :
                    !isNullish(_id) ? { _id: _id } :
                        void 0;
        }
        if (isNullish(id) && !isNullish(_id)) {
            id = _id;
        }
        if (!isNullish(id)) {
            return "".concat(__typename, ":").concat((typeof id === "number" ||
                typeof id === "string") ? id : JSON.stringify(id));
        }
    }
}
var defaultConfig = {
    dataIdFromObject: defaultDataIdFromObject,
    addTypename: true,
    resultCaching: true,
    canonizeResults: false,
};
function normalizeConfig(config) {
    return compact.compact(defaultConfig, config);
}
function shouldCanonizeResults(config) {
    var value = config.canonizeResults;
    return value === void 0 ? defaultConfig.canonizeResults : value;
}
function getTypenameFromStoreObject(store, objectOrReference) {
    return storeUtils.isReference(objectOrReference)
        ? store.get(objectOrReference.__ref, "__typename")
        : objectOrReference && objectOrReference.__typename;
}
var TypeOrFieldNameRegExp = /^[_a-z][_0-9a-z]*/i;
function fieldNameFromStoreName(storeFieldName) {
    var match = storeFieldName.match(TypeOrFieldNameRegExp);
    return match ? match[0] : storeFieldName;
}
function selectionSetMatchesResult(selectionSet, result, variables) {
    if (objects.isNonNullObject(result)) {
        return isArray(result)
            ? result.every(function (item) { return selectionSetMatchesResult(selectionSet, item, variables); })
            : selectionSet.selections.every(function (field) {
                if (storeUtils.isField(field) && directives.shouldInclude(field, variables)) {
                    var key = storeUtils.resultKeyNameFromField(field);
                    return hasOwn.call(result, key) &&
                        (!field.selectionSet ||
                            selectionSetMatchesResult(field.selectionSet, result[key], variables));
                }
                return true;
            });
    }
    return false;
}
function storeValueIsStoreObject(value) {
    return objects.isNonNullObject(value) &&
        !storeUtils.isReference(value) &&
        !isArray(value);
}
function makeProcessedFieldsMerger() {
    return new mergeDeep.DeepMerger;
}
function extractFragmentContext(document, fragments$1) {
    var fragmentMap = fragments.createFragmentMap(getFromAST.getFragmentDefinitions(document));
    return {
        fragmentMap: fragmentMap,
        lookupFragment: function (name) {
            var def = fragmentMap[name];
            if (!def && fragments$1) {
                def = fragments$1.lookup(name);
            }
            return def || null;
        },
    };
}

exports.TypeOrFieldNameRegExp = TypeOrFieldNameRegExp;
exports.defaultDataIdFromObject = defaultDataIdFromObject;
exports.extractFragmentContext = extractFragmentContext;
exports.fieldNameFromStoreName = fieldNameFromStoreName;
exports.getTypenameFromStoreObject = getTypenameFromStoreObject;
exports.hasOwn = hasOwn;
exports.isArray = isArray;
exports.isNullish = isNullish;
exports.makeProcessedFieldsMerger = makeProcessedFieldsMerger;
exports.normalizeConfig = normalizeConfig;
exports.selectionSetMatchesResult = selectionSetMatchesResult;
exports.shouldCanonizeResults = shouldCanonizeResults;
exports.storeValueIsStoreObject = storeValueIsStoreObject;
