// ethers/lib.esm/ethers.js
export { BaseContract, Contract, ContractFactory } from '../../@ethersproject/contracts/lib.esm/index.js';
export { BigNumber } from '../../@ethersproject/bignumber/lib.esm/bignumber.js';
export { FixedNumber } from '../../@ethersproject/bignumber/lib.esm/fixednumber.js';
export { Signer, VoidSigner } from '../../@ethersproject/abstract-signer/lib.esm/index.js';
export { Wallet } from '../../@ethersproject/wallet/lib.esm/index.js';
import * as index$1 from '../../@ethersproject/constants/lib.esm/index.js';
export { index$1 as constants };
import * as index from '../../@ethersproject/providers/lib.esm/index.js';
export { index as providers };
export { getDefaultProvider } from '../../@ethersproject/providers/lib.esm/index.js';
export { Wordlist } from '../../@ethersproject/wordlists/lib.esm/wordlist.js';
export { wordlists } from '../../@ethersproject/wordlists/lib.esm/wordlists.js';
import * as utils from './utils.js';
export { utils };
import { Logger } from '../../@ethersproject/logger/lib.esm/index.js';
export { ErrorCode as errors } from '../../@ethersproject/logger/lib.esm/index.js';
import { version } from './_version.js';
export { version } from './_version.js';

const logger = new Logger(version);

export { logger };
