// @ethersproject/constants/lib.esm/bignumbers.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var bignumber = require('../../bignumber/lib.esm/bignumber.js');
require('../../bignumber/lib.esm/fixednumber.js');

const NegativeOne = ( /*#__PURE__*/bignumber.BigNumber.from(-1));
const Zero = ( /*#__PURE__*/bignumber.BigNumber.from(0));
const One = ( /*#__PURE__*/bignumber.BigNumber.from(1));
const Two = ( /*#__PURE__*/bignumber.BigNumber.from(2));
const WeiPerEther = ( /*#__PURE__*/bignumber.BigNumber.from("1000000000000000000"));
const MaxUint256 = ( /*#__PURE__*/bignumber.BigNumber.from("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"));
const MinInt256 = ( /*#__PURE__*/bignumber.BigNumber.from("-0x8000000000000000000000000000000000000000000000000000000000000000"));
const MaxInt256 = ( /*#__PURE__*/bignumber.BigNumber.from("0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"));

exports.MaxInt256 = MaxInt256;
exports.MaxUint256 = MaxUint256;
exports.MinInt256 = MinInt256;
exports.NegativeOne = NegativeOne;
exports.One = One;
exports.Two = Two;
exports.WeiPerEther = WeiPerEther;
exports.Zero = Zero;
