// @mui/material/Alert/Alert.js
import _objectWithoutPropertiesLoose from '../../../@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js';
import _extends from '../../../@babel/runtime/helpers/esm/extends.js';
import * as React from 'react';
import PropTypes from 'prop-types';
import { clsx } from '../../../clsx/dist/clsx.m.js';
import '../../utils/esm/elementAcceptingRef.js';
import '../../utils/esm/elementTypeAcceptingRef.js';
import 'react-is';
import '../../utils/esm/ponyfillGlobal.js';
import '../../utils/esm/refType.js';
import capitalize from '../../utils/esm/capitalize.js';
import '../../utils/esm/integerPropType.js';
import composeClasses from '../../utils/esm/composeClasses/composeClasses.js';
import '../../base/BadgeUnstyled/BadgeUnstyled.js';
import '../../base/BadgeUnstyled/badgeUnstyledClasses.js';
import '../../base/ButtonUnstyled/ButtonUnstyled.js';
import '../../base/ButtonUnstyled/buttonUnstyledClasses.js';
import '../../base/ClickAwayListener/ClickAwayListener.js';
import '../../base/FocusTrap/FocusTrap.js';
import '../../base/FormControlUnstyled/FormControlUnstyled.js';
import '../../base/FormControlUnstyled/FormControlUnstyledContext.js';
import '../../base/FormControlUnstyled/formControlUnstyledClasses.js';
import '../../base/InputUnstyled/InputUnstyled.js';
import '../../base/InputUnstyled/inputUnstyledClasses.js';
import '../../base/ListboxUnstyled/useListbox.types.js';
import '../../base/MenuUnstyled/MenuUnstyled.js';
import '../../base/MenuUnstyled/MenuUnstyledContext.js';
import '../../base/MenuUnstyled/menuUnstyledClasses.js';
import '../../base/MenuItemUnstyled/MenuItemUnstyled.js';
import '../../base/MenuItemUnstyled/menuItemUnstyledClasses.js';
import '../../base/ModalUnstyled/ModalUnstyled.js';
import '../../base/ModalUnstyled/modalUnstyledClasses.js';
import '../../base/MultiSelectUnstyled/MultiSelectUnstyled.js';
import '../../base/NoSsr/NoSsr.js';
import '../../base/OptionGroupUnstyled/OptionGroupUnstyled.js';
import '../../base/OptionGroupUnstyled/optionGroupUnstyledClasses.js';
import '../../base/OptionUnstyled/OptionUnstyled.js';
import '../../base/OptionUnstyled/optionUnstyledClasses.js';
import '../../base/PopperUnstyled/PopperUnstyled.js';
import '../../base/PopperUnstyled/popperUnstyledClasses.js';
import '../../base/Portal/Portal.js';
import '../../base/SelectUnstyled/SelectUnstyled.js';
import '../../base/SelectUnstyled/selectUnstyledClasses.js';
import '../../base/SliderUnstyled/SliderUnstyled.js';
import '../../base/SliderUnstyled/SliderValueLabelUnstyled.js';
import '../../base/SliderUnstyled/sliderUnstyledClasses.js';
import '../../base/SnackbarUnstyled/SnackbarUnstyled.js';
import '../../base/SnackbarUnstyled/snackbarUnstyledClasses.js';
import '../../base/SwitchUnstyled/SwitchUnstyled.js';
import '../../base/SwitchUnstyled/switchUnstyledClasses.js';
import '../../base/TablePaginationUnstyled/TablePaginationUnstyled.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import '../../base/TablePaginationUnstyled/tablePaginationUnstyledClasses.js';
import '../../base/TabPanelUnstyled/TabPanelUnstyled.js';
import '../../base/TabPanelUnstyled/tabPanelUnstyledClasses.js';
import '../../base/TabsUnstyled/TabsUnstyled.js';
import '../../base/TabsUnstyled/TabsContext.js';
import '../../base/TabsUnstyled/tabsUnstyledClasses.js';
import '../../base/TabsListUnstyled/TabsListUnstyled.js';
import '../../base/TabsListUnstyled/tabsListUnstyledClasses.js';
import '../../base/TabUnstyled/TabUnstyled.js';
import '../../base/TabUnstyled/tabUnstyledClasses.js';
import '../../base/TextareaAutosize/TextareaAutosize.js';
import '@emotion/styled';
import '@emotion/react';
import '../../styled-engine/StyledEngineProvider/StyledEngineProvider.js';
import '../../styled-engine/GlobalStyles/GlobalStyles.js';
import '../../system/esm/borders.js';
import '../../system/esm/display.js';
import '../../system/esm/flexbox.js';
import '../../system/esm/cssGrid.js';
import '../../system/esm/palette.js';
import '../../system/esm/positions.js';
import '../../system/esm/shadows.js';
import '../../system/esm/sizing.js';
import '../../system/esm/spacing.js';
import '../../system/esm/responsivePropType.js';
import '../../system/esm/typography.js';
import '../../system/esm/styleFunctionSx/styleFunctionSx.js';
import '../../system/esm/getThemeValue.js';
import '../../system/esm/Box/Box.js';
import '../../system/esm/useTheme.js';
import '../../system/esm/createStyled.js';
import '../../system/esm/styled.js';
import '../../private-theming/ThemeProvider/ThemeProvider.js';
import '../../private-theming/useTheme/ThemeContext.js';
import { darken, lighten } from '../../system/esm/colorManipulator.js';
import '../../system/esm/ThemeProvider/ThemeProvider.js';
import '../../system/esm/Container/createContainer.js';
import '../../system/esm/Container/Container.js';
import '../../system/esm/Container/containerClasses.js';
import '../../system/esm/Unstable_Grid/Grid.js';
import '../../system/esm/Unstable_Grid/createGrid.js';
import '../../system/esm/Unstable_Grid/gridClasses.js';
import '../../system/esm/Stack/Stack.js';
import '../../system/esm/Stack/createStack.js';
import '../../system/esm/Stack/stackClasses.js';
import styled from '../styles/styled.js';
import useThemeProps from '../styles/useThemeProps.js';
import Paper from '../Paper/Paper.js';
import '../Paper/paperClasses.js';
import alertClasses, { getAlertUtilityClass } from './alertClasses.js';
import IconButton from '../IconButton/IconButton.js';
import '../IconButton/iconButtonClasses.js';
import SuccessOutlinedIcon from '../internal/svg-icons/SuccessOutlined.js';
import ReportProblemOutlinedIcon from '../internal/svg-icons/ReportProblemOutlined.js';
import ErrorOutlineIcon from '../internal/svg-icons/ErrorOutline.js';
import InfoOutlinedIcon from '../internal/svg-icons/InfoOutlined.js';
import ClearIcon from '../internal/svg-icons/Close.js';

const _excluded = ["action", "children", "className", "closeText", "color", "components", "componentsProps", "icon", "iconMapping", "onClose", "role", "severity", "slotProps", "slots", "variant"];
const useUtilityClasses = ownerState => {
  const {
    variant,
    color,
    severity,
    classes
  } = ownerState;
  const slots = {
    root: ['root', `${variant}${capitalize(color || severity)}`, `${variant}`],
    icon: ['icon'],
    message: ['message'],
    action: ['action']
  };
  return composeClasses(slots, getAlertUtilityClass, classes);
};
const AlertRoot = styled(Paper, {
  name: 'MuiAlert',
  slot: 'Root',
  overridesResolver: (props, styles) => {
    const {
      ownerState
    } = props;
    return [styles.root, styles[ownerState.variant], styles[`${ownerState.variant}${capitalize(ownerState.color || ownerState.severity)}`]];
  }
})(({
  theme,
  ownerState
}) => {
  const getColor = theme.palette.mode === 'light' ? darken : lighten;
  const getBackgroundColor = theme.palette.mode === 'light' ? lighten : darken;
  const color = ownerState.color || ownerState.severity;
  return _extends({}, theme.typography.body2, {
    backgroundColor: 'transparent',
    display: 'flex',
    padding: '6px 16px'
  }, color && ownerState.variant === 'standard' && {
    color: theme.vars ? theme.vars.palette.Alert[`${color}Color`] : getColor(theme.palette[color].light, 0.6),
    backgroundColor: theme.vars ? theme.vars.palette.Alert[`${color}StandardBg`] : getBackgroundColor(theme.palette[color].light, 0.9),
    [`& .${alertClasses.icon}`]: theme.vars ? {
      color: theme.vars.palette.Alert[`${color}IconColor`]
    } : {
      color: theme.palette[color].main
    }
  }, color && ownerState.variant === 'outlined' && {
    color: theme.vars ? theme.vars.palette.Alert[`${color}Color`] : getColor(theme.palette[color].light, 0.6),
    border: `1px solid ${(theme.vars || theme).palette[color].light}`,
    [`& .${alertClasses.icon}`]: theme.vars ? {
      color: theme.vars.palette.Alert[`${color}IconColor`]
    } : {
      color: theme.palette[color].main
    }
  }, color && ownerState.variant === 'filled' && _extends({
    fontWeight: theme.typography.fontWeightMedium
  }, theme.vars ? {
    color: theme.vars.palette.Alert[`${color}FilledColor`],
    backgroundColor: theme.vars.palette.Alert[`${color}FilledBg`]
  } : {
    backgroundColor: theme.palette.mode === 'dark' ? theme.palette[color].dark : theme.palette[color].main,
    color: theme.palette.getContrastText(theme.palette[color].main)
  }));
});
const AlertIcon = styled('div', {
  name: 'MuiAlert',
  slot: 'Icon',
  overridesResolver: (props, styles) => styles.icon
})({
  marginRight: 12,
  padding: '7px 0',
  display: 'flex',
  fontSize: 22,
  opacity: 0.9
});
const AlertMessage = styled('div', {
  name: 'MuiAlert',
  slot: 'Message',
  overridesResolver: (props, styles) => styles.message
})({
  padding: '8px 0',
  minWidth: 0,
  overflow: 'auto'
});
const AlertAction = styled('div', {
  name: 'MuiAlert',
  slot: 'Action',
  overridesResolver: (props, styles) => styles.action
})({
  display: 'flex',
  alignItems: 'flex-start',
  padding: '4px 0 0 16px',
  marginLeft: 'auto',
  marginRight: -8
});
const defaultIconMapping = {
  success: /*#__PURE__*/jsx(SuccessOutlinedIcon, {
    fontSize: "inherit"
  }),
  warning: /*#__PURE__*/jsx(ReportProblemOutlinedIcon, {
    fontSize: "inherit"
  }),
  error: /*#__PURE__*/jsx(ErrorOutlineIcon, {
    fontSize: "inherit"
  }),
  info: /*#__PURE__*/jsx(InfoOutlinedIcon, {
    fontSize: "inherit"
  })
};
const Alert = /*#__PURE__*/React.forwardRef(function Alert(inProps, ref) {
  var _ref, _slots$closeButton, _ref2, _slots$closeIcon, _slotProps$closeButto, _slotProps$closeIcon;
  const props = useThemeProps({
    props: inProps,
    name: 'MuiAlert'
  });
  const {
      action,
      children,
      className,
      closeText = 'Close',
      color,
      components = {},
      componentsProps = {},
      icon,
      iconMapping = defaultIconMapping,
      onClose,
      role = 'alert',
      severity = 'success',
      slotProps = {},
      slots = {},
      variant = 'standard'
    } = props,
    other = _objectWithoutPropertiesLoose(props, _excluded);
  const ownerState = _extends({}, props, {
    color,
    severity,
    variant
  });
  const classes = useUtilityClasses(ownerState);
  const AlertCloseButton = (_ref = (_slots$closeButton = slots.closeButton) != null ? _slots$closeButton : components.CloseButton) != null ? _ref : IconButton;
  const AlertCloseIcon = (_ref2 = (_slots$closeIcon = slots.closeIcon) != null ? _slots$closeIcon : components.CloseIcon) != null ? _ref2 : ClearIcon;
  const closeButtonProps = (_slotProps$closeButto = slotProps.closeButton) != null ? _slotProps$closeButto : componentsProps.closeButton;
  const closeIconProps = (_slotProps$closeIcon = slotProps.closeIcon) != null ? _slotProps$closeIcon : componentsProps.closeIcon;
  return /*#__PURE__*/jsxs(AlertRoot, _extends({
    role: role,
    elevation: 0,
    ownerState: ownerState,
    className: clsx(classes.root, className),
    ref: ref
  }, other, {
    children: [icon !== false ? /*#__PURE__*/jsx(AlertIcon, {
      ownerState: ownerState,
      className: classes.icon,
      children: icon || iconMapping[severity] || defaultIconMapping[severity]
    }) : null, /*#__PURE__*/jsx(AlertMessage, {
      ownerState: ownerState,
      className: classes.message,
      children: children
    }), action != null ? /*#__PURE__*/jsx(AlertAction, {
      ownerState: ownerState,
      className: classes.action,
      children: action
    }) : null, action == null && onClose ? /*#__PURE__*/jsx(AlertAction, {
      ownerState: ownerState,
      className: classes.action,
      children: /*#__PURE__*/jsx(AlertCloseButton, _extends({
        size: "small",
        "aria-label": closeText,
        title: closeText,
        color: "inherit",
        onClick: onClose
      }, closeButtonProps, {
        children: /*#__PURE__*/jsx(AlertCloseIcon, _extends({
          fontSize: "small"
        }, closeIconProps))
      }))
    }) : null]
  }));
});
process.env.NODE_ENV !== "production" ? Alert.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // |     To update them edit the d.ts file and run "yarn proptypes"     |
  // ----------------------------------------------------------------------
  /**
   * The action to display. It renders after the message, at the end of the alert.
   */
  action: PropTypes.node,
  /**
   * The content of the component.
   */
  children: PropTypes.node,
  /**
   * Override or extend the styles applied to the component.
   */
  classes: PropTypes.object,
  /**
   * @ignore
   */
  className: PropTypes.string,
  /**
   * Override the default label for the *close popup* icon button.
   *
   * For localization purposes, you can use the provided [translations](/material-ui/guides/localization/).
   * @default 'Close'
   */
  closeText: PropTypes.string,
  /**
   * The color of the component. Unless provided, the value is taken from the `severity` prop.
   * It supports both default and custom theme colors, which can be added as shown in the
   * [palette customization guide](https://mui.com/material-ui/customization/palette/#adding-new-colors).
   */
  color: PropTypes /* @typescript-to-proptypes-ignore */.oneOfType([PropTypes.oneOf(['error', 'info', 'success', 'warning']), PropTypes.string]),
  /**
   * The components used for each slot inside.
   *
   * This prop is an alias for the `slots` prop.
   * It's recommended to use the `slots` prop instead.
   *
   * @default {}
   */
  components: PropTypes.shape({
    CloseButton: PropTypes.elementType,
    CloseIcon: PropTypes.elementType
  }),
  /**
   * The extra props for the slot components.
   * You can override the existing props or add new ones.
   *
   * This prop is an alias for the `slotProps` prop.
   * It's recommended to use the `slotProps` prop instead, as `componentsProps` will be deprecated in the future.
   *
   * @default {}
   */
  componentsProps: PropTypes.shape({
    closeButton: PropTypes.object,
    closeIcon: PropTypes.object
  }),
  /**
   * Override the icon displayed before the children.
   * Unless provided, the icon is mapped to the value of the `severity` prop.
   * Set to `false` to remove the `icon`.
   */
  icon: PropTypes.node,
  /**
   * The component maps the `severity` prop to a range of different icons,
   * for instance success to `<SuccessOutlined>`.
   * If you wish to change this mapping, you can provide your own.
   * Alternatively, you can use the `icon` prop to override the icon displayed.
   */
  iconMapping: PropTypes.shape({
    error: PropTypes.node,
    info: PropTypes.node,
    success: PropTypes.node,
    warning: PropTypes.node
  }),
  /**
   * Callback fired when the component requests to be closed.
   * When provided and no `action` prop is set, a close icon button is displayed that triggers the callback when clicked.
   * @param {React.SyntheticEvent} event The event source of the callback.
   */
  onClose: PropTypes.func,
  /**
   * The ARIA role attribute of the element.
   * @default 'alert'
   */
  role: PropTypes.string,
  /**
   * The severity of the alert. This defines the color and icon used.
   * @default 'success'
   */
  severity: PropTypes.oneOf(['error', 'info', 'success', 'warning']),
  /**
   * The extra props for the slot components.
   * You can override the existing props or add new ones.
   *
   * This prop is an alias for the `componentsProps` prop, which will be deprecated in the future.
   *
   * @default {}
   */
  slotProps: PropTypes.shape({
    closeButton: PropTypes.object,
    closeIcon: PropTypes.object
  }),
  /**
   * The components used for each slot inside.
   *
   * This prop is an alias for the `components` prop, which will be deprecated in the future.
   *
   * @default {}
   */
  slots: PropTypes.shape({
    closeButton: PropTypes.elementType,
    closeIcon: PropTypes.elementType
  }),
  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.bool])), PropTypes.func, PropTypes.object]),
  /**
   * The variant to use.
   * @default 'standard'
   */
  variant: PropTypes /* @typescript-to-proptypes-ignore */.oneOfType([PropTypes.oneOf(['filled', 'outlined', 'standard']), PropTypes.string])
} : void 0;

export { Alert as default };
