// ethers/lib.esm/ethers.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index$3 = require('../../@ethersproject/contracts/lib.esm/index.js');
var bignumber = require('../../@ethersproject/bignumber/lib.esm/bignumber.js');
var fixednumber = require('../../@ethersproject/bignumber/lib.esm/fixednumber.js');
var index = require('../../@ethersproject/abstract-signer/lib.esm/index.js');
var index$1 = require('../../@ethersproject/wallet/lib.esm/index.js');
var index$4 = require('../../@ethersproject/constants/lib.esm/index.js');
var index$2 = require('../../@ethersproject/providers/lib.esm/index.js');
var wordlist = require('../../@ethersproject/wordlists/lib.esm/wordlist.js');
var wordlists = require('../../@ethersproject/wordlists/lib.esm/wordlists.js');
var utils = require('./utils.js');
var index$5 = require('../../@ethersproject/logger/lib.esm/index.js');
var _version = require('./_version.js');

const logger = new index$5.Logger(_version.version);

exports.BaseContract = index$3.BaseContract;
exports.Contract = index$3.Contract;
exports.ContractFactory = index$3.ContractFactory;
exports.BigNumber = bignumber.BigNumber;
exports.FixedNumber = fixednumber.FixedNumber;
exports.Signer = index.Signer;
exports.VoidSigner = index.VoidSigner;
exports.Wallet = index$1.Wallet;
exports.constants = index$4;
exports.getDefaultProvider = index$2.getDefaultProvider;
exports.providers = index$2;
exports.Wordlist = wordlist.Wordlist;
exports.wordlists = wordlists.wordlists;
exports.utils = utils;
Object.defineProperty(exports, 'errors', {
	enumerable: true,
	get: function () { return index$5.ErrorCode; }
});
exports.version = _version.version;
exports.logger = logger;
