// @ethersproject/json-wallets/lib.esm/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var crowdsale = require('./crowdsale.js');
var inspect = require('./inspect.js');
var keystore = require('./keystore.js');

function decryptJsonWallet(json, password, progressCallback) {
    if (inspect.isCrowdsaleWallet(json)) {
        if (progressCallback) {
            progressCallback(0);
        }
        const account = crowdsale.decrypt(json, password);
        if (progressCallback) {
            progressCallback(1);
        }
        return Promise.resolve(account);
    }
    if (inspect.isKeystoreWallet(json)) {
        return keystore.decrypt(json, password, progressCallback);
    }
    return Promise.reject(new Error("invalid JSON wallet"));
}
function decryptJsonWalletSync(json, password) {
    if (inspect.isCrowdsaleWallet(json)) {
        return crowdsale.decrypt(json, password);
    }
    if (inspect.isKeystoreWallet(json)) {
        return keystore.decryptSync(json, password);
    }
    throw new Error("invalid JSON wallet");
}

exports.decryptCrowdsale = crowdsale.decrypt;
exports.getJsonWalletAddress = inspect.getJsonWalletAddress;
exports.isCrowdsaleWallet = inspect.isCrowdsaleWallet;
exports.isKeystoreWallet = inspect.isKeystoreWallet;
exports.decryptKeystore = keystore.decrypt;
exports.decryptKeystoreSync = keystore.decryptSync;
exports.encryptKeystore = keystore.encrypt;
exports.decryptJsonWallet = decryptJsonWallet;
exports.decryptJsonWalletSync = decryptJsonWalletSync;
